generator/findby.go

```diff
@@ -0,0 +1,192 @@
+		}
+	}
+}
+
+func writeTplCode(buf *bytes.Buffer, parent *Model, f *Field, tpl string) {
+	validType, err := validType(f)
+	if err != nil {
+		return
+	}
+
+	property := f.Name
+	query := parent.QueryName
+	model := parent.Name
+	buf.WriteString(fmt.Sprintf(tpl, property, query, validType, model))
+}
+
+// validType returns the short name of the type that can be used to search for
+// the passed Field, in an autogenerated 'FindBy';
+// if the Field is not valid for an autogenerated 'FindBy' it will return an error
+func validType(f *Field) (string, error) {
+	analycedType := f.Node.Type()
+	collectionAlreadyScanned := false
+	for {
+		valid, deepest := lookupValidRecursively(f.Node.Pkg(), analycedType)
+		if valid != nil {
+			return getShortName(analycedType), nil
+		}
+
+		if collectionAlreadyScanned {
+			break
+		}
+
+		singular := singularType(deepest)
+		if singular == nil {
+			break
+		}
+
+		analycedType = singular
+		collectionAlreadyScanned = true
+	}
+
+	return "", fmt.Errorf("kallax: '%s' is not valid for FindBy", f.Node.Type())
+}
+
+// lookupValid returns the first valid type looking into the underlying types of
+// the passed one; if no valid found it will return the deepest underlying
+func lookupValidRecursively(pkg *types.Package, typ types.Type) (valid, deepest types.Type) {
+	for _, isValid := range typeValidators {
+		if isValid(pkg, typ) {
+			return typ, nil
+		}
+	}
+
+	underlaying := typ.Underlying()
+	switch {
+	case typ != underlaying:
+		return lookupValidRecursively(pkg, underlaying)
+	default:
+		return nil, typ
+	}
+}
+
+// singularType returns the type of the elements of the passed collection,
+// or nill if the passed type is not a collection
+func singularType(typ types.Type) types.Type {
+	switch typ.(type) {
+	case *types.Array:
+		return typ.(*types.Array).Elem()
+	case *types.Slice:
+		return typ.(*types.Slice).Elem()
+	}
+
+	return nil
+}
+
+// isBasicType returns true if passed type is one of the followings:
+// string, byte, bool, float(s), int(s) or uint(s)
+func isBasicType(pkg *types.Package, typ types.Type) bool {
+	s := typ.String()
+	return s == "string" || s == "bool" || s == "byte" || s == "float64" || s == "float32" ||
+		s == "int" || s == "int8" || s == "int16" || s == "int32" || s == "int64" ||
+		s == "uint" || s == "uint8" || s == "uint16" || s == "uint32" || s == "uint64"
+}
+
+// isSpecialType returns true if the passed type is one of the kallaxKnownTypes
+// or a types.SQLType
+func isSpecialType(pkg *types.Package, typ types.Type) bool {
+	_, ok := kallaxKnownTypeName(typ)
+	return ok || isSQLType(pkg, typ)
+}
+
+func kallaxKnownTypeName(typ types.Type) (string, bool) {
+	special, ok := kallaxKnownTypes[cleanName(typ, false)]
+	return special, ok
+}
+
+func getShortName(typ types.Type) string {
+	if special, ok := kallaxKnownTypeName(typ); ok {
+		return special
+	}
+
+	return cleanName(typ, true)
+}
+
+func cleanName(typ types.Type, requireShortName bool) string {
```

Comment:
this is already implemented https://github.com/src-d/go-kallax/blob/master/generator/types.go#L697